src/mds/MDCache.cc

```diff
@@ -9292,6 +9292,26 @@ void MDCache::request_forward(MDRequestRef& mdr, mds_rank_t who, int port)
     dout(7) << "request_forward " << *mdr << " to mds." << who << " req "
             << *mdr->client_request << dendl;
     mds->forward_message_mds(mdr->release_client_request(), who);
+    if (mdr->is_batch_head) {
+      BatchOpContext* fin = NULL;
+      int mask = mdr->client_request->head.args.getattr.mask;
+
+      if (mdr->client_request->get_op() == CEPH_MDS_OP_GETATTR) {
+	CInode* in = mdr->in[0];
+	if (in) {
+	  fin = in->batch_ops[mask];
+	}
+      } else {
+	CDentry* dn = mdr->dn[0].back();
+	if (dn) {
+	  fin = dn->batch_ops[mask];
+	}
+      }
+      if (fin) {
+	fin->set_forward_target(who);
+	fin->complete(0);
+      }
```

Comment:
in->batch_ops[mask] and in->batch_ops[mask] leak?

Reply:
@ukernel Hi, thanks for your review:-)

Reply:
@ukernel Hi, thanks for your review:-)
It seems that both in->batch_ops[mask] and dn->batch_ops[mask] would be deleted in the Context::complete method. Am I right?

Reply:
I don't find corresponding code.  I found 
````
      if (mdr->client_request->get_op() == CEPH_MDS_OP_GETATTR) {
        mdr->in[0]->batch_ops.erase(mask);
      } else {
        mdr->dn[0].back()->batch_ops.erase(mask);
      }

````
in Server::respond_to_request. I think we should have similar code here.

Reply:
@ukernel Hi, the code is in "src/include/Context.h":-)
```
class Context {
  Context(const Context& other);
  const Context& operator=(const Context& other);

 protected:
  virtual void finish(int r) = 0;

  // variant of finish that is safe to call "synchronously."  override should
  // return true.
  virtual bool sync_finish(int r) {
    return false;
  }

 public:
  Context() {}
  virtual ~Context() {}       // we want a virtual destructor!!!
  virtual void complete(int r) {
    finish(r);
    **delete this;**
  }
  virtual bool sync_complete(int r) {
    if (sync_finish(r)) {
      delete this;
      return true;
    }
    return false;
  }
};

```

Reply:
It deletes itself. but mdr->in[0]->batch_ops[mask] is pointing to a freed context